Added static Tiberian Sun low bridges

Mailaender opened the pull request:
A slightly improved and rebased version of https://github.com/OpenRA/OpenRA/pull/10764 which now imports dead bridges as dead actors that are rendered transparent in the map editor.

Mailaender committed:
Add static Tiberian Sun low bridges.

Mailaender committed:
Reimport map "A River Runs Near It"

Mailaender committed:
Reimport map "River Raid"

pchote commented:
This doesn't even try to address the main issue that makes #10764 unacceptable. I've gone to great lengths to explain why that approach is no good, and you even appeared to understand it.  I don't understand why you've reopened virtually the same thing again.

:-1: because, as discussed earlier, all of these changes would need to be reverted and rewritten from scratch by the person who wants to implement bridge damage and repair.

Mailaender commented:
>  all of these changes would need to be reverted and rewritten from scratch by the person who wants to implement bridge damage and repair.

Actually not if this get written with bridge damage and repair done via actor replacement.

pchote commented:
I have already explained why that is a horrible solution that I consider to be completely unacceptable.

pchote commented:
Please replace the individual overlay types (remember that TS did not use actors for bridges) with bridge actors that define proper damage states.

abcdefg30 commented on OpenRA.Mods.TS/Traits/Buildings/LowBridge.cs:
```diff
@@ -0,0 +1,55 @@
+			var cells = FootprintUtils.PathableTiles(self.Info.Name, buildingInfo, self.Location);
+			var tileSet = self.World.Map.Rules.TileSet;
+			foreach (var cell in cells)
+				self.World.Map.CustomTerrain[cell] = tileSet.GetTerrainIndex(info.TerrainType);
```
Couldn't you move the `tileSet.GetTerrainIndex(info.TerrainType);` out of the loop?
E.g.
```
var terrainIndex = tileSet.GetTerrainIndex(info.TerrainType);
foreach (var cell in cells)
	self.World.Map.CustomTerrain[cell] = terrainIndex;
```

Mailaender replied:
Yes. Done.

Mailaender commented:
I already tried that https://github.com/Mailaender/OpenRA/commit/d21988f4fc102dc5d2f1f88f2ce7da754441e632. It ends in a rubbish solution.

abcdefg30 commented on mods/ts/sequences/bridges.yaml:
```diff
@@ -0,0 +1,92 @@
+
+lobrdg27:
+	idle: lobrdg01
+		Start: 1
```
Won't inheriting `lobrdg01` and just overwriting the file used for the idle sequence (like already done above) work as well?

Mailaender replied:
Can you give me an example how it should work?

abcdefg30 replied:
Sure.
```
lobrdg27:
	Inherits: lobrdg01
	idle: lobrdg01
```
(Or maybe `Defaults: lobrdg01` instead of `idle: lobrdg01`, didn't try it myself yet.)

Mailaender replied:
This is obsolete now with multiple bridge actors combined.

Mailaender commented:
Rebased so this can be conveniently tested with https://github.com/OpenRA/OpenRA/pull/11211. Replaced individual bridge actors with combined ones importing their supposed damage state. This looks a bit bland as we don't support sprite image randomization, but it isn't terrible.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -72,40 +72,43 @@ class ImportTSMapCommand : IUtilityCommand
+			{ 0x61, "lobrdg_r_ne" }, // lobrdg24
+			{ 0x62, "lobrdg_r_sw" }, // lobrdg25
+			{ 0x63, "lobrdg_r_sw" }, // lobrdg26
+			{ 0x64, "lobrdg_b" }, // lobrdg27
```
My spec from #8916 was to have these two dead versions as their own actor types.  This (1) lets us handle the special case adjacency rendering that only the dead state needs, and (2) will let us eventually remove the hacky zombie / resurrection logic that is now only used for bridges.

Mailaender replied:
Added the dead husk actors back.

Mailaender commented:
This is now looking pretty slick with the transparent bridge husks in the editor:

![image](https://cloud.githubusercontent.com/assets/756669/15992755/5f0eee98-30d4-11e6-8481-858c4a0ef3c6.png)

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -121,6 +124,51 @@ class ImportTSMapCommand : IUtilityCommand
+			{ 0x63, DamageState.Heavy },
+
+			// placeholders for resurrection
+			{ 0x64, DamageState.Dead },
```
These should be imported as Undamaged now.

Mailaender replied:
Removed the zombie bridges.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -303,7 +351,26 @@ static void ReadOverlay(Map map, IniFile file, int2 fullSize)
+							else if (damageState == DamageState.Undamaged)
+								health = 100;
+
+							ar.Add(new HealthInit(health, true));
```
This should be skipped if `health == 100`.

Mailaender replied:
Done.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -72,40 +72,43 @@ class ImportTSMapCommand : IUtilityCommand
+			{ 0x4E, "lobrdg_b_nw" }, // lobrdg05
+			{ 0x4F, "lobrdg_b_se" }, // lobrdg06
+			{ 0x50, "lobrdg_b" }, // lobrdg07
+			{ 0x51, "lobrdg_b_nw" }, // lobrdg08
```
I'm pretty certain that these two are the art states for the dead bridge (your `lobrdg_b_d`) when there is an intact bridge neighbour (because they are not a full cell, so cannot be walkable).

pchote replied:
Scratch that, after inspecting the original game I see that these should actually be converted to `lobrdg_b`.

Mailaender replied:
They are not duplicates. If you stitch the tiles together with an undamaged one in the editor using https://github.com/OpenRA/OpenRA/pull/10764 you can see that they leave a gap at the northern or southern side, which is why I have to differentiate them. It may be easier to see with the `lobrdg_a`  variant, because it has a larger gap.

pchote replied:
See my comment in the main thread for more context on how this works.

Mailaender replied:
Shall I set them up as different facings of the same actor?

pchote replied:
Facings isn't a good fit.  This will want to use directional checks like walls.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -72,40 +72,43 @@ class ImportTSMapCommand : IUtilityCommand
+			{ 0x57, "lobrdg_a_sw" }, // lobrdg14
+			{ 0x58, "lobrdg_a_ne" }, // lobrdg15
+			{ 0x59, "lobrdg_a" }, // lobrdg16
+			{ 0x5A, "lobrdg_a_sw" }, // lobrdg17
```
And likewise these two are the alternative art states for `lobrdg_b_d` when there are intact neighbours.

pchote replied:
And these to `lobrdg_a`.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -121,6 +124,51 @@ class ImportTSMapCommand : IUtilityCommand
+
+			{ 0x50, DamageState.Heavy },
+
+			{ 0x51, DamageState.Critical },
```
And then these should go back to DamageState.Undamaged to match.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -121,6 +124,51 @@ class ImportTSMapCommand : IUtilityCommand
+
+			{ 0x59, DamageState.Heavy },
+
+			{ 0x5A, DamageState.Critical },
```
and these two too.

pchote commented on OpenRA.Mods.TS/UtilityCommands/ImportTSMapCommand.cs:
```diff
@@ -72,40 +72,43 @@ class ImportTSMapCommand : IUtilityCommand
+			{ 0x64, "lobrdg_b_d" }, // lobrdg27
+			{ 0x65, "lobrdg_a_d" }, // lobrdg28
+
+			/* High bridge segments
```
The art frames of these appear to be exact duplicates of the other undamaged ramps.  They definitely aren't for high bridges.

pchote commented:
I tested the original game in detail and see now that it selects the neighbour-based artwork *on the living* actors rather than the dead.  This is the opposite of how the earlier games did it, but the same rendering principles apply on our side.

pchote commented:
The real issue here is matching the art frames to defined (I suspect pairs of own, neighbour) damage state.  We can't define oddballs like `lobrdg_b_se` because that would mean that those artwork states couldn't be used by the standard bridge types when they or their neighbours are damaged.

Mailaender commented:
Updated this with your recent findings.

abcdefg30 commented:
Looks good to me. :+1: / :white_check_mark:

pchote commented:
Needs a rebase.

Mailaender commented:
Rebased.

pchote commented on OpenRA.Mods.TS/Traits/Buildings/LowBridge.cs:
```diff
@@ -0,0 +1,50 @@
+			this.info = info;
+		}
+
+		void INotifyCreated.Created(Actor self)
```
This will need to be done on AddedToWorld, and then a matching RemovedFromWorld should remove the custom terrain.

I'm not sure why I missed this earlier, sorry.

Mailaender replied:
Not yet, the bridges are indestructible for the moment.

pchote replied:
Sure, but that's no excuse to leave this logic half-unfinished in the meantime.

Mailaender replied:
It adds code with no test case, which will probably be broken then.

pchote replied:
The code to avoid bogus state on removal is not complicated.  It can be tested once by making the bridges destroyable and checking with your terrain debug layers.

Mailaender replied:
You are forcing me to jump ahead far from what I wanted to do in this initial pull request. I added destructible bridges and added a test case that we can keep as it won't be game breaking on that map.

Mailaender committed:
Reimport Tread Lightly

Mailaender committed:
Add a destroyable low bridge to Tread Lightly.

pchote commented on OpenRA.Mods.TS/Traits/Buildings/LowBridge.cs:
```diff
@@ -24,27 +25,39 @@ class LowBridgeInfo : ITraitInfo, Requires<BuildingInfo>
+			UpdateTerrain(self, terrainIndex);
+		}
+
+		void INotifyRemovedFromWorld.RemovedFromWorld(Actor self)
```
You'll need to trigger a check (probably in a FrameEndTask) to kill any actors that are no longer allowed in the cell.  Otherwise you end up with orphaned units where they don't belong:

![screen shot 2016-08-26 at 16 59 58](https://cloud.githubusercontent.com/assets/167819/18011871/d9f2a310-6bae-11e6-8c77-39e3b0674aaa.png)

pchote replied:
Note that my last request was to tidy up a specific code issue with the trait so that we could merge the code at a logical and self-contained stopping point.  The extra changes that you added on top of this exposes a new problem which you could alternatively avoid here by removing them for a future PR.  So please don't go off on me about scope-creep, I'm just trying to avoid merging things that are obviously half-baked.

Mailaender replied:
This is a fairly simple and reasonable request that is well inside my skill scope so I don't mind.

Mailaender commented:
Also updated the map importer to avoid adding 3x more bridges than required. It works fine except for some `lobrdg_a` actors. The reason is probably me still not understanding the isometric coordinate system and how to traverse it in an orderly manner.

pchote committed:
Add bridge depth metadata.

pchote committed:
Fix duplicate import of multi-cell overlays.

pchote commented:
Please reset this onto https://github.com/pchote/OpenRA/commits/lowbridges and then we can get this merged.

Mailaender commented:
Thanks for helping me out with the importer. It now imports bridges without the need for manual tweaking. :+1:

abcdefg30 commented:
Looks still good to me. :+1: / :white_check_mark:

pchote commented:
:tada:

Mailaender commented:
:bridge_at_night:
